Add K3s module

joebowbeer opened the pull request:
Adds K3s module. Fixes #863 

This is WIP. In particular, more tests and examples are needed, but so far it seems to work.

Notes:
1. `@types/dockerode` is missing a HostConfig field (and an older version of `dockerode` is being used, besides). I don't know why `HostConfig.CgroupnsMode` needs to be configured, but the Go and .Net versions of K3s container do it.
1. The Java and Go versions of K3s container run k3s with a `--tls-san` option. If this is required, I'll need to figure out how to obtain the container host before it is started.
1. The k3s-container is depending on the async iterator support added in tar-stream@3. This version is currently hoisted to the top-level in the package-lock, but its position may be tenuous.

cristianrgreco commented:
Thanks @joebowbeer! Do you mind making this PR a draft if it's not ready for review/merge?

joebowbeer commented:
@cristianrgreco I converted this to a draft. I _think_ it will be ready after I add more tests and doc examples.

If you can scan the notes in the description and let me know if I'm missing something, I'd appreciate it.

joebowbeer commented:
The failure is  unrelated to my changes?

```
FAIL packages/testcontainers/src/generic-container/generic-container.test.ts (81.142 s)
  ● GenericContainer › should return first mapped port

    (HTTP code 500) server error - Head "https://registry-1.docker.io/v2/cristianrgreco/testcontainer/manifests/1.1.14": Get "https://auth.docker.io/token?account=githubactions&scope=repository%3Acristianrgreco%2Ftestcontainer%3Apull&service=registry.docker.io": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

      at ../../node_modules/docker-modem/lib/modem.js:343:17
      at IncomingMessage.<anonymous> (../../node_modules/docker-modem/lib/modem.js:370:9)
```

joebowbeer commented:
@cristianrgreco I converted to draft until issues are resolved. See details and questions below.

Passing tests:

https://github.com/testcontainers/testcontainers-node/actions/runs/11814284520/job/32951551243?pr=864

    Docker (rootful?)
    external host 172.17.0.3


Failing tests:

https://github.com/testcontainers/testcontainers-node/actions/runs/11814284520/job/32951558708?pr=864

    Podman (rootless)
    external host 10.88.0.3

I suspect it is the rootless podman that is preventing k3s from running, as k3s runs `withPrivilegedMode`, and is the first node module to do so.

According to limitations listed in the Java docs

https://java.testcontainers.org/modules/k3s/#known-limitations

> K3sContainer runs as a privileged container and needs to be able to spawn its own containers. For these reasons, K3sContainer will not work in certain rootless Docker, Docker-in-Docker, or other environments where privileged containers are disallowed.

The other potential cause is the [missing --tls-san host](https://github.com/testcontainers/testcontainers-go/pull/1289), but if so I expect that problem won't manifest until privileged containers are permitted.

Thoughts?

joebowbeer commented:
@cristianrgreco I made the following changes:
1. Added warning about rootful requirement to docs
2. Disabled tests that run K3sContainer if CI_ROOTLESS
3. Added getAliasedKubeConfig for running in custom network

The tests pass for me on macos with Rancher Desktop and with Podman (rootful).

joebowbeer commented:
FYI - Leaving a bread crumb in case `/sys/fs/cgroup` [bind mount](https://github.com/testcontainers/testcontainers-node/blob/fa96946ec5f64d789dfad320e9cd6493c43042d3/packages/modules/k3s/src/k3s-container.ts#L20) proves problematic.

From https://github.com/k3s-io/k3s/issues/4873#issuecomment-1007282314

> Docker Desktop now uses cgroupv2. If you need to run systemd in a container then:
> The following options are needed: --privileged --cgroupns=host -v /sys/fs/cgroup:/sys/fs/cgroup:rw.

joebowbeer commented:
> @joebowbeer If you think there's an issue with the module, then a PR is welcome

Understood. I'm not aware of an issue at the moment; only a diff between the implementations that I'm hoping to understand, eventually.
